File: examples/part2-pyramid-py/signup.py
"""Python companion for the Part 2 lesson "The Testing Pyramid". The same
feature is implemented in Rust, Go, and Python, with tests at all three levels:
unit (pure validator), integration (service + real in-memory adapters), and
end-to-end (the App composition root).

Run the tests:

    python3 -m unittest discover -s . -p 'test_*.py'
"""

from __future__ import annotations

from typing import Protocol

# ---------------------------------------------------------------- validation


class InvalidEmailError(Exception):
    """The email failed validation."""


class DuplicateEmailError(Exception):
    """The email is already registered."""


def validate_email(email: str) -> None:
    """Pure validation — the base of the pyramid: no I/O, instant tests.

    >>> validate_email("ada@example.com")  # no exception means valid
    """
    if not email:
        raise InvalidEmailError("email is empty")
    if any(c.isspace() for c in email):
        raise InvalidEmailError("email contains whitespace")
    parts = email.split("@")
    if len(parts) != 2 or not parts[0] or not parts[1] or "." not in parts[1]:
        raise InvalidEmailError("email format is invalid")


# ------------------------------------------------------------------- service


class UserRepo(Protocol):
    """The storage port."""

    def exists(self, email: str) -> bool: ...
    def save(self, email: str) -> None: ...


class Notifier(Protocol):
    """The notification port."""

    def send_welcome(self, email: str) -> None: ...


class InMemoryRepo:
    """The storage adapter used in tests and this demo."""

    def __init__(self) -> None:
        self._emails: set[str] = set()

    def exists(self, email: str) -> bool:
        return email in self._emails

    def save(self, email: str) -> None:
        self._emails.add(email)


class RecordingNotifier:
    """Records welcomes (a real one would talk SMTP)."""

    def __init__(self) -> None:
        self.sent: list[str] = []

    def send_welcome(self, email: str) -> None:
        self.sent.append(email)


class SignupService:
    """The middle of the pyramid: logic coordinating the two ports."""

    def __init__(self, repo: UserRepo, notifier: Notifier) -> None:
        self._repo = repo
        self._notifier = notifier

    def signup(self, email: str) -> None:
        validate_email(email)
        if self._repo.exists(email):
            raise DuplicateEmailError(f"{email} already registered")
        self._repo.save(email)
        self._notifier.send_welcome(email)


# ----------------------------------------------------------------------- app


class App:
    """The top of the pyramid: the composition root a CLI/web layer would call,
    returning the user-visible message.
    """

    def __init__(self) -> None:
        self._service = SignupService(InMemoryRepo(), RecordingNotifier())

    def signup(self, email: str) -> str:
        try:
            self._service.signup(email)
        except DuplicateEmailError:
            return f"{email} is already registered."
        except InvalidEmailError:
            return f"'{email}' is not a valid email address."
        return f"Welcome, {email}! Check your inbox."

File: examples/part2-pyramid-go/signup.go
// Package signup is the Go companion for the Part 2 lesson "The Testing
// Pyramid". The same feature is implemented in Rust, Go, and Python, with
// tests at all three levels: unit (pure validator), integration (service +
// real in-memory adapters), and end-to-end (the App composition root).
//
//	go test ./...
package signup

import (
	"errors"
	"fmt"
	"strings"
)

// ---------------------------------------------------------------- validation

var (
	ErrEmpty      = errors.New("email is empty")
	ErrBadFormat  = errors.New("email format is invalid")
	ErrDuplicate  = errors.New("email already registered")
	ErrWhitespace = errors.New("email contains whitespace")
)

// ValidateEmail is pure — the base of the pyramid: no I/O, instant tests.
func ValidateEmail(email string) error {
	if email == "" {
		return ErrEmpty
	}
	if strings.ContainsAny(email, " \t\n") {
		return ErrWhitespace
	}
	parts := strings.Split(email, "@")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" || !strings.Contains(parts[1], ".") {
		return ErrBadFormat
	}
	return nil
}

// ------------------------------------------------------------------- service

// UserRepo is the storage port.
type UserRepo interface {
	Exists(email string) bool
	Save(email string)
}

// Notifier is the notification port.
type Notifier interface {
	SendWelcome(email string)
}

// InMemoryRepo is the storage adapter used in tests and this demo.
type InMemoryRepo struct{ emails map[string]bool }

func NewInMemoryRepo() *InMemoryRepo { return &InMemoryRepo{emails: make(map[string]bool)} }

func (r *InMemoryRepo) Exists(email string) bool { return r.emails[email] }
func (r *InMemoryRepo) Save(email string)        { r.emails[email] = true }

// RecordingNotifier records welcomes (a real one would talk SMTP).
type RecordingNotifier struct{ Sent []string }

func (n *RecordingNotifier) SendWelcome(email string) { n.Sent = append(n.Sent, email) }

// Service is the middle of the pyramid: logic coordinating the two ports.
type Service struct {
	repo     UserRepo
	notifier Notifier
}

func NewService(repo UserRepo, notifier Notifier) *Service {
	return &Service{repo: repo, notifier: notifier}
}

func (s *Service) Signup(email string) error {
	if err := ValidateEmail(email); err != nil {
		return err
	}
	if s.repo.Exists(email) {
		return fmt.Errorf("%s: %w", email, ErrDuplicate)
	}
	s.repo.Save(email)
	s.notifier.SendWelcome(email)
	return nil
}

// ----------------------------------------------------------------------- app

// App is the top of the pyramid: the composition root a binary would call,
// returning the user-visible message.
type App struct{ service *Service }

func NewApp() *App {
	return &App{service: NewService(NewInMemoryRepo(), &RecordingNotifier{})}
}

func (a *App) Signup(email string) string {
	err := a.service.Signup(email)
	switch {
	case err == nil:
		return fmt.Sprintf("Welcome, %s! Check your inbox.", email)
	case errors.Is(err, ErrDuplicate):
		return fmt.Sprintf("%s is already registered.", email)
	default:
		return fmt.Sprintf("'%s' is not a valid email address.", email)
	}
}
